test/1.4/misc/linemarks.dml

```diff
@@ -0,0 +1,218 @@
+/*
+  © 2023 Intel Corporation
+  SPDX-License-Identifier: MPL-2.0
+*/
+
+dml 1.4;
+
+device test;
+
+/// COMPILE-ONLY
+
+// The whitespace before %} is important to make that line get its dedicated
+// line in generated C, instead of it being omitted
+// BEGIN-LINEMARKS-CBLOCK
+header %{
+
+
+ %}
+// END-LINEMARKS
+
+
+// BEGIN-LINEMARKS-CBLOCK
+footer %{
+
+
+
+ %}
+// END-LINEMARKS
+
+template u {
+    shared method s_m() default
+    // BEGIN-LINEMARKS-FUNC _DML_TM_u__s_m
+    {
+    }
+    // END-LINEMARKS
+}
+
+template t is u {
+    // Downcasting done by shared method overrides should not receive linemarks
+    shared method s_m()
+    // BEGIN-LINEMARKS-FUNC _DML_TM_t__s_m
+    {
+    }
+    // END-LINEMARKS
+
+    // The prelude of memoized startup methods should have no linemarks
+    shared independent startup memoized method s_memo_startup()
+        -> (int, bool, float)
+    // BEGIN-LINEMARKS-FUNC _DML_TM_t__s_memo_startup
+    {
+        return (4, true, 4.7);
+    }
+    // END-LINEMARKS
+}
+
+independent startup memoized method memo_startup() -> (int, bool, float)
+// BEGIN-LINEMARKS-FUNC _DML_M_memo_startup
+{
+    return (4, true, 4.7);
+}
+// END-LINEMARKS
+
+// no linemarks
+saved int sav_1;
+
+// BEGIN-LINEMARKS-FUNC _init_data_objs
+saved int sav_2 = 4;
+// END-LINEMARKS
+
+method init()
+// BEGIN-LINEMARKS-FUNC _DML_M_init
+{
+    // Compound.toc() omits braces when there are no declarations
+    // NOT-LINEMARKED
+    {
+        assert true;
+        assert true;
+    // NOT-LINEMARKED
+    }
+
+    // .toc() of empty compounds and null statements emits no C statements
+    // NOT-LINEMARKED
+    {}
+    // NOT-LINEMARKED
+    ;
+    if (true)
+    // mkCompound reduces a compound with a single statement to that statement
+    // NOT-LINEMARKED
+    {
+        assert true;
+    // NOT-LINEMARKED
+    }
+
+    // ... Unless it contains a declaration.
+    {
+        local int y;
+    }
+
```

Comment:
As discussed elsewhere, the way we use NOT-LINEMARKED to test how compound statement collapse is somewhat unfocused and smokey, and may possibly cause unnecessary disruption. Ideally it might be expressed as a ctree_test, but not today. Still, these tests are of some use so discarding them right away would be a waste -- how about moving the less important tests out to a separate method, with a comment on this status? That will make it easier to take proper action if they ever get in the way.

Reply:
Done.